### ISSUE TITLE
SPIKE: distinguish between `conda.*_prefixes`

### ISSUE THREAD

**Issue by kenodegard** [conda/conda#12704]
### What happened?

The conda context defines the following prefixes:

1. `conda_prefix`: this is where conda is installed, `sys.prefix`
2. `root_prefix`: this can be defined in `.condarc`, otherwise it defaults to the `conda_prefix`
3. `active_prefix`: this is the currently active environment, `$CONDA_PREFIX`
4. `default_prefix`: this is `active_prefix` if defined, otherwise it defaults to the `base` environment (the `base` environment name can be controlled via `$CONDA_DEFAULT_ENV`)
5. `target_prefix`: this is the environment specified on the CLI via `--name`/`--prefix`, default to `default_conda`

In the vast majority of use cases we expect:
1. `conda_prefix` == `root_prefix`
6. for subcommands where `--name`/`--prefix` is not provided (e.g., `conda info`)
    - `active_prefix` == `default_prefix` == `target_prefix`
3. for subcommands where `--name`/`--prefix` is provided (e.g., `conda create`) 
    - `active_prefix` == `default_prefix`
    - `target_prefix` == `--name`/`--prefix`

These many prefixes are inconsistently (and possibly incorrectly used) throughout conda.

Furthermore, the way in which the context object is initialized also sets `$CONDA_PREFIX` when running select subcommands (those that define `--name`/`--prefix`), this in particular makes resetting the context object (for testing/mocking purposes) especially difficult:

https://github.com/conda/conda/blob/72977501b6c3bc6573ec0e9ffce630fed64229cd/conda/base/context.py#L446-L468

**Comment by kenodegard** [comment]
Some history:

- `conda_prefix`
    - introduced in 4.2.0
    - see https://github.com/conda/conda/commit/6e730cc2f7c992df61d067243a9421f4732453a5
    - part of the private conda environment effort
- `root_prefix` (previously `conda.config.root_dir` → `conda.config.ROOT_DIR`)
    - introduced in 1.1.0
    - see https://github.com/conda/conda/commit/d2b370a705de8969252158fb4a483ce0ab1eeb28
    - originally intended for testing purposes only
- `default_prefix` (previously `conda.config.default_prefix` → `conda.config.DEFAULT_ENV_PREFIX` → `conda.config.default_environment` → `conda.anaconda.default_environment`)
    - introduced in 1.1.0
    - see https://github.com/conda/conda/commit/acd8144f0a2b3af4164396311350a053ea160069
    - part of a larger configuration overhaul
- `active_prefix`
    - introduced in 4.4.0
    - see https://github.com/conda/conda/commit/cbbe81f38b7feb85757c3b21c9c7ccd15d978d66
    - part of cleaning up the `conda info` output
- `target_prefix` (previously `prefix` → `conda.cli.common.get_prefix()` → `conda.cli.utils.get_prefix()`)
    - introduced in 1.3.0
    - see https://github.com/conda/conda/commit/9fd0f4b49dd08e0437d5cd091e22ac1faca75637
    - to abstract out the `--name`/`--prefix` logic from the CLI implementations
- context's `$CONDA_PREFIX` initialization
    - introduced in 4.4.0
    - see https://github.com/conda/conda/commit/00b8d2de226c994355c0353436331d729b3aaa3e
    - addressing incorrect condarc loading for subcommands with `--name`/`--prefix`

**Comment by travishathaway** [comment]
@kenodegard,

I think the assumptions you make about the various "prefix" values is fair.

Are we now going to make a proposal for changing this? Or should we just make sure that our documentation about using these context object properties is clearer?

I actually had to rely on the various meanings of the "prefix" variables when I recently created this plugin:

- https://github.com/travishathaway/conda-envlock/

I'd be curious to see if I actually used these correctly and whether that might serve as a good example of how to and how not to use the "prefix" variables.
